src/nxt_http_js.c

```diff
@@ -165,8 +167,9 @@ nxt_http_js_ext_get_args(njs_vm_t *vm, njs_object_prop_t *prop,
 
     args = njs_value_arg(&val);
 
-    ret = njs_vm_query_string_parse(vm, r->args->start,
-                                    r->args->start + r->args->length, args);
+    length = r->args->length;
+    start = (length != 0) ? r->args->start : (u_char *) "";
```

Comment:
length can't be < 0. so I'd make it (length > 0)

Reply:
But forget that, that's the wrong thing to check, we should check if _r->args->start_ is NULL.

Again, if it even can be?

Reply:
> Again, if it even can be?

Yes, relevant test and Sanitizer's backtrace can be found in commit message to this patch. As an additional verification I can suggest to add assert in code:

```
diff --git a/src/nxt_http_js.c b/src/nxt_http_js.c
--- a/src/nxt_http_js.c
+++ b/src/nxt_http_js.c
@@ -165,6 +165,8 @@ nxt_http_js_ext_get_args(njs_vm_t *vm, n
 
     args = njs_value_arg(&val);
 
+    nxt_assert(r->args->start != NULL);
+
     ret = njs_vm_query_string_parse(vm, r->args->start,
                                     r->args->start + r->args->length, args);
 
```

And run test to make sure:

```
% pytest test/test_rewrite.py::test_rewrite_njs | grep "assertion"
2024/02/01 11:12:03.190 [alert] 85574#19767557 src/nxt_http_js.c:168 assertion failed: r->args->start != NULL
```

Reply:
OK, well, I don't know what tool had found these issues and whether they were theoretical or it actually hit them... so good to know these are real...

Reply:
tests are available here https://github.com/nginx/unit/tree/master/test and can be run as any standard pytests (i.e. `pytest test/test_static.py`). [UndefinedBehaviorSanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) can be enabled just by compiling Unit with `-fsanitize=undefined` option.

Reply:
As a different approach, maybe we could just avoid `.start` being `NULL` at all, and set it to `(u_char *) ""` whenever length is 0.

That would need modifying `nxt_null_string()` and `nxt_str_null()`, and then fixing other code that is using NULL to use "".

The benefit of this approach is simplicity: less states a nxt_str_t can have.

Although it would need some research.

Reply:
Currently, there are two states of `nxt_str_t` that mean the same thing (or I haven't seen any code making a meaningful difference from them):

```c
{.length = 0, .start = NULL}
{.length = 0, .start = ""}
```

Reply:
Oh, now I remember, some strings are NULL due to some zero allocation, so it's not even set.  Forget my suggestion.  :)